Use boot file metadata instead of maintaining a separate map

bhagany opened the pull request:
The most immediate benefit of this change is that `markdown` no longer needs to manually track files removed from the fileset between invocations. As a result, `markdown` is shorter and simpler, and the same will be true of any future input parsing tasks.

Some discussion is necessary on this one. There is a problem about how to deal with requiring `boot` in `io.perun.core`. The problem arises when a project doesn't depend on boot itself, and then executes a fn in a pod, and the fn's namespace requires `io.perun.core`. The pod can't find boot.core, and so, boom. Options:
 - Make projects depend on boot
 - Add boot to pod dependencies when pod namespaces require io.perun.core
 - Split the boot-using parts of io.perun.core into a separate namespace, and don't require that namespace from a pod
 - Something else?

I prefer the second option, but not strongly, and I'm not sure if there are performance considerations with adding more dependencies to pods.

bhagany committed:
Merge branch 'trace' into boot-file-meta

bhagany committed:
Use boot file metadata instead of maintaining a separate map
The most immediate benefit of this change is that `markdown` no longer
needs to manually track files removed from the fileset between
invocations. As a result, `markdown` is shorter and simpler, and
the same will be true of any future input parsing tasks.

bhagany committed:
Update docstring

bhagany committed:
merge consistent-filterers -> boot-file-meta

bhagany committed:
merge consistent-filterers -> boot-file-meta

bhagany committed:
Merge branch 'master' into boot-file-meta

bhagany committed:
fix docstring typo

podviaznikov commented on src/io/perun/core.clj:
```diff
@@ -11,15 +12,15 @@
    Internally it's stored as a map indexed by `:path`"
   [fileset]
-  (-> fileset meta +meta-key+ vals))
+  (keep +meta-key+ (vals (:tree fileset))))
```
what is keep function?

bhagany replied:
`(keep :key coll)` is like `(filter identity #(map :key coll))` https://clojuredocs.org/clojure.core/keep

bhagany replied:
Perhaps I should explain - `get-meta` returns data in the same shape as before - a seq of metadata maps.  This PR doesn't have any API changes, aside from not needing to deal with removed markdown files.

podviaznikov commented on src/io/perun/core.clj:
```diff
@@ -11,15 +12,15 @@
+  "Update `+meta-key+` metadata for files in `data` and return updated fileset"
   [fileset data]
-  (vary-meta fileset assoc +meta-key+ (key-meta data)))
+  (boot/add-meta fileset (into {} (for [d data] [(:path d) {+meta-key+ d}]))))
```
is this new boot function or it was there for a long time?

bhagany replied:
I had to look it up, I am pretty new to boot.  Here is the commit that added it https://github.com/boot-clj/boot/commit/1ebcbb8dd78ee4fb88dad70e629630f9442a8b7a.  I think this means it has been in boot since 2.0?

bhagany replied:
actually, on a second look, I think the function was just moved in this commit.  Here is the real commit: https://github.com/boot-clj/boot/commit/038d8dde09704e3d22b5cdfdd8be984825f04bf9

Deraen commented:
> Split the boot-using parts of io.perun.core into a separate namespace, and don't require that namespace from a pod

Is probably the correct solution. It is generally best to presume that `boot.core` is not available in pods.

bhagany committed:
Move perun meta handling to separate namespace
Because metadata handling is now using `boot.core`, and
`io.perun.core` is required from pods, a separate namespace is
necessary to avoid having every pod depend on `boot.core` being
available.

bhagany commented:
Okay @Deraen, sounds good to me.  I've committed that change.

podviaznikov commented:
Looks good to me. Merging.
